dftimewolf/lib/processors/gcp_cloud_resource_tree_helper.py

```diff
@@ -0,0 +1,303 @@
+      indented based on it's location in the tree.
+    """
+    tab = '\t'
+    output: List[Any] = []
+
+    # Count the number of parent resources and the level in the tree the
+    # resource is at.
+    level = 0
+    parent_resource = self.parent
+    while parent_resource:
+      level = level + 1
+      parent_resource = parent_resource.parent
+
+    # Add the resource parent entries to the List of dictionaries
+    counter = level
+    parent_resource = self.parent
+    if parent_resource:
+      while counter > 0:
+        counter = counter - 1
+        entry: Dict[str, Any] = {}
+        entry['resource_object'] = parent_resource
+        entry['graph'] = f'{tab*counter}|--{parent_resource.name}'
+        output.insert(0, entry)
+        if parent_resource.parent:
+          parent_resource = parent_resource.parent
+
+    # Add resource entry to the List of dictionaries
+    entry = {}
+    entry['resource_object'] = self
+    entry['graph'] = f'{tab*level}|--{self.name}'
+    output.insert(level, entry)
+
+    # Add resource children entries to the List of dictionaries
+    output.extend(self.GenerateChildrenTree(level + 1))
+
+    return output
+
+  def GenerateChildrenTree(self, level: int) -> List[Dict[str, 'Resource']]:
+    """Generates the resource children tree.
+
+    Args:
+      level: The level in the tree to place the children at.
+    Returns:
+      List of dictionaries containing a reference to the children resource and
+          their names indented based on their location in the tree.
+    """
+    result: List[Dict[str, Resource]] = []
+    tab = '    '
+
+    for child in self.children:
+      entry: Dict[str, Any] = {}
+      entry['resource_object'] = child
+      entry['graph'] = f'{tab*level}|--{child.name}'
+      result.append(entry)
+
+      if child.children:
+        result.extend(child.GenerateChildrenTree(level + 1))
+
+    return result
+
+  def __str__(self) -> str:
+    """Returns a string representation of the resource tree."""
+    output = '\n'
+    dashes = '-' * 200
+
+    # Draw table header
+    output = output + dashes + '\n'
+    # pylint: disable=line-too-long
+    output = output + '{:<20s}|{:<18s}|{:<20s}|{:<25s}|{:<18s}|{:<20s}|{:<25s}|{:<18s}|{:<100s}\n'.format(
+        'Resource ID', 'Resource Type', 'Creation TimeStamp', 'Created By',
+        'Creator IP Addr', 'Deletion Timestamp', 'Deleted By',
+        'Deleter IP Addr', 'Tree')
+    output = output + dashes + '\n'
+
+    result = self.GenerateTree()
+    for i in result:
+      resource: Optional[Resource] = i.get('resource_object')
+      if resource:
+        output = output + \
+            ('{:<20s}|{:<18s}|{:<20s}|{:<25s}|{:<18s}|{:<20s}|{:<25s}|{:<18s}|{:<100s}\n'.format( resource.id if ('-' not in resource.id) else "" , resource.type, resource.creation_timestamp.astimezone(timezone.utc).strftime("%Y-%m-%d %H:%M:%S") if resource.creation_timestamp else "",
```

Comment:
as you use this string more than once, you can make it a local variable and use it here and in line 260, it will also help shorten the line
'{:<20s}|{:<18s}|{:<20s}|{:<25s}|{:<18s}|{:<20s}|{:<25s}|{:<18s}|{:<100s}\n'

Reply:
Done